Add http server for serving type files during compilation to Federated Types

RussellCanfield opened the pull request:
These changes target use cases where you may be running multiple remotes locally in a monorepo scenario. Currently you can run into two scenarios:

- Timeouts waiting for a remote to start - Remote A that requires types from Remote B, you start webpack serve on both remotes together.
- Dead locks when Remote A has a bi-directional share with Remote B, you start webpack serve on both remotes together. Webpack Dev Server middleware prevents serving until the compilation is complete.

There are 2 new configuration options available for this plugin:

## TypeFetchOptions
Fetch Options represent configuring a simple retry mechanism for fetching remote types. These are on by default, worst case this results in a bad URL/non-existent host being retries several times. There are clear fetch/axios logs in this scenario (same as before). The user can turn off checking for non-existing types using **shouldRetryOnTypesNotFound**. This is on by default because in the first condition above, there are no types available yet.

```typescript
export interface TypeFetchOptions {
  /** The maximum time to wait for downloading remote types in milliseconds.
   * @default 2000  */
  downloadRemoteTypesTimeout?: number;
  /** The maximum number of retry attempts.
   * @default 3  */
  maxRetryAttempts?: number;
  /** The default number of milliseconds between retries.
   * @default 1000  */
  retryDelay?: number;
  /** Should retry if no types are found in destination.
   * @default true  */
  shouldRetryOnTypesNotFound?: boolean;
  /** Should retry type fetching operations.
   * @default true  */
  shouldRetry?: boolean;
}
```

## TypeServeOptions
ServeOptions allow you to host the types outside of webpack-dev-server and this is off by default. In the bi-directional example above, we need the ability to move them outside of webpack-dev-server since its a blocker.

```typescript
export interface TypeServeOptions {
  /** The port to serve type files on, this is separate from the webpack dev server port. */
  port?: number;
  /** The host to serve type files on. Example: 'localhost' */
  host?: string;
}
```

Together both options allow you to successfully bypass these dead-lock scenarios. The one note here is that there is an assumption on when it's a "serve" vs a "build". Webpack seems to have some deprecated options (ex: watch, watchmode, etc), but on a build we stop the type server if using it. On a serve we keep it running to support re-compiles, otherwise you get failures after the initial compilation.

This change also improves parsing of the remote entry from the federationConfig. Previously this relied on having the remote entry "remote@http://localhost/" at a bare minimum - the key being the trailing slash. Now it will handle without the trailing slash or with. This makes more sense when you are hosting the types separately since there isn't a "remoteEntry.js"

RussellCanfield commented:
@ScriptedAlchemy - this should be ready to go, there are some odd edge cases between NX testing and my regular pnpm workspace monorepo. There are 3 sort of dirty workarounds:

- isServe flag, I set this if the watch hook runs for the main reason of, if you run a build on compiler.done we need to close node. If its serve and we are watching we need to keep it open otherwise it tries to download types and fails.

- The server.ts file, needed a map because in NX it seemed to run multiple instances of this script (maybe require cache?), the map fixed it versus a simple server variable storing the reference.

- Type needs to be generated at 2 lifecycles, beforeCompile so there are available at download - download happens beforeCompile which seems fitting? But if you set clean output in webpack, again afterEmit. I think this is because the types are webpack "assets" but it's a guess. There may be a better longterm solution here but the type generation is pretty fast unless you are exposing like 100 components.

brunos3d commented on packages/typescript/src/lib/server.ts:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    const server = http.createServer((req, res) => {
+      const { url: fileName } = req;
+
+      if (!fileName) {
+        logger.log(`Unable to find file: ${fileName} in request url`);
+        res.end();
+        return;
+      }
+
+      const safeSuffix = path
+        .normalize(fileName)
+        .replace(/^(\.\.(\/|\\|$))+/, '');
+      const safeJoin = path.join(outputPath, safeSuffix);
+
+      logger.log(`Retrieving file: ${safeJoin}`);
+
+      const fileStream = fs.createReadStream(safeJoin);
+
+      fileStream.on('error', (err) => {
+        logger.log(`Error reading file: ${err}`);
+        res.statusCode = 500;
+        res.end();
+      });
+
+      res.writeHead(200, { 'Content-Type': 'text/plain' });
+      fileStream.pipe(res);
```
I found out some vulnerabilities here like:
- Directory Traversal Attacks: code is using path.normalize and path.join to build the file path, but it can be easily hacked if you put some more back steps to the relative path, e.g: `http://example.com:8080/../../../../../etc/passwd`.
- Error Handling: if something unexpected occurs during the file serving, or even trying to find it, the request will continue open, we need to use some defensive tactics here.

Here is a simple suggestion on how to solve it

```suggestion
      const fileName = path.join(outputPath, req.url);

      try {
        // Ensure the requested file is within the specified directory
        if (!fileName.startsWith(outputPath)) {
          res.writeHead(403, { 'Content-Type': 'text/plain' });
          res.end('Forbidden');
          return;
        }

        // Check if the file exists
        fs.stat(fileName, (err, stat) => {
          if (err) {
            logger.log(`Error reading file: ${err}`);
            res.writeHead(500, { 'Content-Type': 'text/plain' });
            res.end('Internal Server Error');
          } else {
            if (stat.isFile()) {
              res.writeHead(200, { 'Content-Type': 'text/plain' });
              fs.createReadStream(fileName).pipe(res);
            } else {
              // Handle non-file requests (e.g., directories)
              res.writeHead(404, { 'Content-Type': 'text/plain' });
              res.end('Not Found');
            }
          }
        });
      } catch (err) {
        logger.log(`Error reading file: ${err}`);
        res.writeHead(500, { 'Content-Type': 'text/plain' });
        res.end('Internal Server Error');
      }
```

brunos3d replied:
If we are looking to use it in production, it will need to have some more options, but if its only for dev envs, I believe the error handling suggestion is fine by now :smile:

RussellCanfield commented:
@brunos3d - thanks for the review! I went ahead and merged that suggestion, seems CodeQL flagged it so I added back the normalize call and I think it hasn't re-scanned. But this was a good suggestion.

I retested in both of my monorepos (pnpm and nx), and everything seems to be working. I added one small tweak to avoid downloading types twice, it seems "beforeCompile" can be called multiple times so this just resulted in silent errors and additional attempts

brunos3d commented:
Awesome, well done @RussellCanfield! :handshake:

brunos3d reviewed (approved):
LGTM :rocket:

sherbakovdev commented:
Unfortunately, the solution doesn't completely fix the deadlock with bi-directional sharing. I'm using Rspack, and the plugin blocks the compilation for both shared remotes. When plugin timeouts, the compilation finishes.

Basically, both remotes are getting stuck on "Getting types index for remote" and will never be compiled while plugin is running.

RussellCanfield commented:
> Unfortunately, the solution doesn't completely fix the deadlock with bi-directional sharing. I'm using Rspack, and the plugin blocks the compilation for both shared remotes. When plugin timeouts, the compilation finishes.
> 
> Basically, both remotes are getting stuck on "Getting types index for remote" and will never be compiled while plugin is running.

Did you try using the config options for serving types on another port? If so I would suggest migrating to the native federated types package which I believe works for rspack. It should be easy to migrate to

ilteoood commented:
Hi @sherbakovdev, since you are using rspack can you give a shot to the native-federation-typescript plugin?

sherbakovdev commented:
Wow! Thanks for the swift replies, folks! I appreciate it! 

I'm pretty new to MF and didn't know native-federation-typescript exists. I am going to try now.

Thanks again!

sherbakovdev commented:
It seems the native plugin works—there is still a minor issue with the race condition though.

`
frontends/remote dev: Rspack compiled successfully in 2.23 s
frontends/remote dev: Federated types created correctly
services/shell dev: Rspack compiled successfully in 2.31 s

frontends/remote dev: Unable to download federated types for 'shell' from 'http://localhost:3000/@mf-types.zip' because 'Request failed with status code 404', skipping...

frontends/remote dev: Federated types extraction completed
services/shell dev: Federated types extraction completed
services/shell dev: Federated types created correctly
`

ilteoood commented:
Could create a repro and share it with us? I'll take a look at it and try to improve the plugin.

sherbakovdev commented:
Here you go @ilteoood https://github.com/sherbakovdev/mf-types-demo

ilteoood commented:
I don't think we can fix this scenario. MF works because components are requested at runtime, while types are needed during build time.

What I could do is implement a configurable retry logic: it will still happen, but at least it could recover by itself

@sherbakovdev

sherbakovdev commented:
That sounds like a good solution! Thanks!

ilteoood commented:
@sherbakovdev this should be available now with version 0.3.0 of both native-federation-tests and native-federation-typescript packages
